drivers/regulator/regulator_npm6001.c

```diff
@@ -0,0 +1,702 @@
+#define NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM_POS 3
+#define NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM                                   \
+	BIT(NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM_POS)
+
+#define NPM6001_BUCKMODEPADCONF_BUCKMODE0PADTYPE_CMOS  BIT(0)
+#define NPM6001_BUCKMODEPADCONF_BUCKMODE1PADTYPE_CMOS  BIT(1)
+#define NPM6001_BUCKMODEPADCONF_BUCKMODE2PADTYPE_CMOS  BIT(2)
+#define NPM6001_BUCKMODEPADCONF_BUCKMODE0PULLD_ENABLED BIT(4)
+#define NPM6001_BUCKMODEPADCONF_BUCKMODE1PULLD_ENABLED BIT(5)
+#define NPM6001_BUCKMODEPADCONF_BUCKMODE2PULLD_ENABLED BIT(6)
+
+/* nPM6001 PADDRIVESTRENGTH fields */
+#define NPM6001_PADDRIVESTRENGTH_READY_HIGH BIT(2)
+#define NPM6001_PADDRIVESTRENGTH_NINT_HIGH  BIT(3)
+#define NPM6001_PADDRIVESTRENGTH_SDA_HIGH   BIT(5)
+
+struct regulator_npm6001_pconfig {
+	struct i2c_dt_spec i2c;
+	uint8_t buck_pad_val;
+	uint8_t pad_val;
+};
+
+struct regulator_npm6001_config {
+	struct regulator_common_config common;
+	const struct device *p;
+	uint8_t source;
+};
+
+struct regulator_npm6001_data {
+	struct regulator_common_data data;
+};
+
+static const struct linear_range buck0_range =
+	LINEAR_RANGE_INIT(1800000, 100000U, 0x0U, 0xFU);
+
+static const struct linear_range buck1_range =
+	LINEAR_RANGE_INIT(700000, 50000U, 0x0U, 0xEU);
+
+static const struct linear_range buck2_range =
+	LINEAR_RANGE_INIT(1200000, 50000U, 0xAU, 0xEU);
+
+static const struct linear_range buck3_range =
+	LINEAR_RANGE_INIT(500000, 25000U, 0x0U, 0x70U);
+
+static int regulator_npm6001_ldo0_list_voltage(const struct device *dev,
+					       unsigned int idx,
+					       int32_t *volt_uv)
+{
+	switch (idx) {
+	case 0U:
+		*volt_uv = 1800000;
+		break;
+	case 1U:
+		*volt_uv = 2100000;
+		break;
+	case 2U:
+		*volt_uv = 2410000;
+		break;
+	case 3U:
+		*volt_uv = 2700000;
+		break;
+	case 4U:
+		*volt_uv = 3000000;
+		break;
+	case 5U:
+		*volt_uv = 3300000;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int regulator_npm6001_buck012_set_voltage(
+	const struct device *dev, int32_t min_uv, int32_t max_uv,
+	const struct linear_range *range, uint8_t vout_reg, uint8_t conf_reg)
+{
+	const struct regulator_npm6001_config *config = dev->config;
+	const struct regulator_npm6001_pconfig *pconfig = config->p->config;
+	uint8_t conf;
+	uint16_t idx;
+	int ret;
+
+	ret = linear_range_get_win_index(range, min_uv, max_uv, &idx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* force PWM mode when updating voltage */
+	ret = i2c_reg_read_byte_dt(&pconfig->i2c, conf_reg, &conf);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if ((conf & NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM) != 0U) {
+		ret = i2c_reg_write_byte_dt(
+			&pconfig->i2c, conf_reg,
+			conf | NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	ret = i2c_reg_write_byte_dt(&pconfig->i2c, vout_reg, (uint8_t)idx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = i2c_reg_write_byte_dt(&pconfig->i2c, vout_reg + 1U, (uint8_t)idx);
```

Comment:
I need to check, I took https://github.com/NordicSemiconductor/pmic-npm6001-lib/blob/main/src/lib_npm6001.c#L86 as a reference, where they seem to do individual writes.

Reply:
I'm observing a weird behavior when using `i2c_write_dt` (if I for example do in batch force PWM/update voltage/disable PWM the BUCK voltage jumps up to VIN). Let's leave this for now.

Reply:
I can confirm the same behavior with BUCK0/1/2, in those voltage is not updated. I guess the I2C stop condition is needed somehow.

Reply:
It can be that the register address is not incremented, so one cannot access multiple registers in one transaction (unfortunately, the PS does not precise that). Let's stay with what works.

Reply:
ok, managed to use `i2c_burst_write_dt` for vout successfully, updated